### PR TITLE
modified the failure report and added codecoverage report when enabled

### PR DESCRIPTION
**Pull request by peetkes**
format junit is modified so it will return nr otf tests, success and fails per testcase.
Also returns coverage info per testcase

Example:
```
<testsuite errors="0" failures="0" hostname="localhost" name="contexts/cds/components/security/test-suite" tests="18" time="8.193104" timestamp="2023-05-09T12:38:10.945312+02:00">
    <testcase classname="test-authenticate.xqy" name="test-authenticate.xqy" time="0.074011" tests="10" succes="10" fail="0">
        <coverage module="/contexts/cds/components/common/lib/config.xqy">
            <wanted count="232">38 76 117 143 162 163 176 177 178 191 192 205 206 207 220 221 234 235 236 237 238 251 252 265 281 283 284 286 288 289 291 292 309 324 339 353 366 379 392 405 418 431 444 457 470 483 501 514 516 517 531 544 557 570 582 583 584 586 596 598 599 600 612 623 634 645 656 671 673 675 679 681 684 686 687 705 706 708 710 712 713 719 721 722 740 741 755 756 757 758 772 774 775 790 791 792 794 795 797 799 800 802 804 819 821 822 834 835 849 852 853 866 868 870 872 873 875 878 880 882 884 885 886 887 888 889 890 911 912 915 916 918 919 920 924 925 927 930 932 934 937 938 940 943 945 947 948 950 951 956 958 960 962 963 969 971 972 973 1006 1007 1008 1010 1037 1038 1039 1041 1043 1045 1048 1050 1052 1053 1055 1057 1078 1080 1081 1095 1096 1098 1100 1101 1103 1107 1108 1110 1114 1115 1117 1119 1121 1122 1124 1125 1126 1127 1139 1154 1156 1158 1159 1160 1161 1162 1164 1165 1166 1179 1180 1182 1185 1187 1199 1211 1213 1215 1227 1228 1230 1231 1232 1233 1234 1235 1247 1248 1250 1251 1252 1253 1254 1255</wanted>
            <covered count="0"/>
            <missing count="232">38 76 117 143 162 163 176 177 178 191 192 205 206 207 220 221 234 235 236 237 238 251 252 265 281 283 284 286 288 289 291 292 309 324 339 353 366 379 392 405 418 431 444 457 470 483 501 514 516 517 531 544 557 570 582 583 584 586 596 598 599 600 612 623 634 645 656 671 673 675 679 681 684 686 687 705 706 708 710 712 713 719 721 722 740 741 755 756 757 758 772 774 775 790 791 792 794 795 797 799 800 802 804 819 821 822 834 835 849 852 853 866 868 870 872 873 875 878 880 882 884 885 886 887 888 889 890 911 912 915 916 918 919 920 924 925 927 930 932 934 937 938 940 943 945 947 948 950 951 956 958 960 962 963 969 971 972 973 1006 1007 1008 1010 1037 1038 1039 1041 1043 1045 1048 1050 1052 1053 1055 1057 1078 1080 1081 1095 1096 1098 1100 1101 1103 1107 1108 1110 1114 1115 1117 1119 1121 1122 1124 1125 1126 1127 1139 1154 1156 1158 1159 1160 1161 1162 1164 1165 1166 1179 1180 1182 1185 1187 1199 1211 1213 1215 1227 1228 1230 1231 1232 1233 1234 1235 1247 1248 1250 1251 1252 1253 1254 1255</missing>
        </coverage>
...
```

### CODE REVIEW

**Comment by rjrudin** [comment]
I'm trying this out now - can you squash the 2 commits into a single commit first?

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestCase.java [hunk 1]
```diff
@@ -49,11 +71,31 @@ public double getTime() {
...
 
 	public void setTestFailures(List<JUnitTestFailure> testFailures) {
 		this.testFailures = testFailures;
 	}
+
+  public List<JUnitTestCoverage> getTestCoverages() {
+    return testCoverages;
+  }
+
+  public void setTestCoverages(List<JUnitTestCoverage> testCoverages) {
```
Comment (rjrudin) [inline marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestCase.java:98]: Remove this as it's not needed (it looks like `setTestFailures` isn't needed either, but let's keep that as-is to minimize the changes in this PR).

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestCoverage.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      this.coverageInfo = new ArrayList<>();
+    }
+    this.coverageInfo.add(coverageInfo);
+  }
+
+  public List<JUnitTestCoverageInfo> getCoverageInfo() {
+    return coverageInfo;
+  }
+
+  public void setCoverageInfo(List<JUnitTestCoverageInfo> coverageInfo) {
```
Comment (rjrudin) [inline marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestCoverage.java:30]: Remove this, not needed.

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestSuite.java [hunk 3]
```diff
@@ -9,23 +9,26 @@
...
-	public JUnitTestSuite(String xml, int errors, int failures, String hostname, String name, int tests, double time) {
+	public JUnitTestSuite(String xml, int errors, int failures, String hostname, String name, int cases, int tests, double time) {
 		this.xml = xml;
 		this.errors = errors;
 		this.failures = failures;
 		this.hostname = hostname;
 		this.name = name;
 		this.tests = tests;
 		this.time = time;
+    this.cases = cases;
```
Comment (rjrudin) [inline marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestSuite.java:31]: What is the difference between `tests` and `cases`? I'm thinking first that `cases` should be `testCases`, but looking at this, I don't know what the difference between the two is.

**Review by rjrudin — CHANGES_REQUESTED** [review]
Thanks for submitting - there are more comments than usual here as we're doing design-via-PR, so I have some questions about the overall design in here.

**Comment by rjrudin** [comment]
One other thought - I think this will require an enhancement in ml-gradle's UnitTestTask so that there's an option to see coverage data - or that just gets printed by default.

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestSuite.java (hunk 3 above)
Comment (peetkes) [reply marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestSuite.java:31]: Tests is all tests, cases is all testcases, cases should be less than tests

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestCase.java (hunk 1 above)
Comment (peetkes) [reply marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestCase.java:98]: removed

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/DefaultJUnitTestReporter.java [hunk 4]
```diff
@@ -13,7 +15,7 @@ public String reportOnJUnitTestSuites(List<JUnitTestSuite> suites) {
 
 		int testsCompleted = 0;
 		int testFailed = 0;
-
+    sb.append("\n\nFailure Report:");
```
[the remark this answers is not in the document]
Comment (peetkes) [reply marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/DefaultJUnitTestReporter.java:18]: Changed

File: marklogic-unit-test-modules/src/main/ml-modules/root/test/test-coverage.xqy [hunk 5]
```diff
@@ -213,17 +214,21 @@ declare function cover:results(
     $results-map as map:map,
     $results as item()*)
 {
+  (:)  xdmp:trace($TRACE-ID, $results-map),:)
```
[the remark this answers is not in the document]
Comment (peetkes) [reply marklogic-unit-test-modules/src/main/ml-modules/root/test/test-coverage.xqy:217]: removed

File: marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestSuite.java (hunk 3 above)
Comment (rjrudin) [reply marklogic-unit-test-client/src/main/java/com/marklogic/test/unit/JUnitTestSuite.java:31]: We likely can't change it in a minor release, but should `tests` really be `testSuites` then? And cases is `testCases`? 

My understanding of marklogic-unit-test is there are two terms when referring to "tests":

1. A "test" refers to a file that becomes a module in the modules database, and it can have many assertions in it - e.g. `test:assert-equal`. 
2. A "suite" refers to a directory that contains one to many tests. 

I don't know of a term that captures "all suites" though. 

So basically, I'm wondering what "cases" refers to. Is it something different than a "test" or a "suite", or is it the same as one of those?

**Comment by rjrudin** [comment]
@peetkes I tried this out in ml-gradle via the following draft PR - 

The output doesn't look so good, and it's not due to this PR - I'm just thinking this coverage feature is very half-baked and should be fixed before it's exposed any further:

```
Coverage Results:

Coverage Per Test

example-failing-tests.xqy > example-failing-tests.xqy COVERED
    [module: /example-lib.xqy, coverageInfo: [[wanted: 3], [covered: 0], [missing: 3]]]
    [module: /test/test-helper.xqy, coverageInfo: [[wanted: 336], [covered: 0], [missing: 336]]]

example-lib-tests.xqy > example-lib-tests.xqy COVERED
    [module: /example-lib.xqy, coverageInfo: [[wanted: 3], [covered: 2], [missing: 1]]]
    [module: /test/test-helper.xqy, coverageInfo: [[wanted: 336], [covered: 9], [missing: 327]]]

example-tests.sjs > example-tests.sjs COVERED
    [module: /example-lib.xqy, coverageInfo: [[wanted: 3], [covered: 2], [missing: 1]]]
    [module: /test/test-helper.xqy, coverageInfo: [[wanted: 336], [covered: 9], [missing: 327]]]

3 tests covered

Coverage Across All Tests:
        [module: /example-lib.xqy, coverageInfo: {wanted=3, covered=2, missing=1}]
        [module: /test/test-helper.xqy, coverageInfo: {wanted=336, covered=9, missing=327}]

```

Issues that I see (which are also visible in the GUI test runner):

1. SJS modules are ignored.
2. The coverage on example-lib.xqy is incorrect - it should state that 5 lines were wanted and 3 were missed.
3. The user doesn't care about test-helper.xqy, and its inclusion throws off the stats

The GUI is arguably worse, as it includes other modules that the user doesn't care about:

image

Are you getting better results with this on your own project? The problem for me is that while this project is not yet supported, ml-gradle is, and I don't want to further expose this coverage feature in ml-gradle (currently, you can include `-PrunCodeCoverage`, but the results only appear in the XML test results file).

**Comment by hansenmc** [comment]
The line counts are obtained from stepping through the debugger breakpoints. 
https://github.com/marklogic-community/marklogic-unit-test/blob/master/marklogic-unit-test-modules/src/main/ml-modules/root/test/test-coverage.xqy#L107

Not every line in the file will be represented in the line count, and some multi-line statements may be seen as a single line. 

For the modules that the user doesn't care about. The challenge is to figure out which modules the user doesn't care about. It picks up the code that is in-scope, which includes built-in modules and some of the lib modules for the test coverage, but what if you were evaluating code coverage of the code coverage lib? Or of the built-in MarkLogic modules?

**Comment by rjrudin** [comment]
Closing for now until I can dig deeper into the existing coverage ability and determine where to go with it.
